Add support for accelerating mean and rms with numpy

huffman opened the pull request:
The previous implementation used Python's built-in ** operator and math.sum to calculate the mean and rms values of an AnalogSpan.

On my machine (Ryzen 5950x), this took around 11s to run all measurements on 30M samples. VoltageMeasurer, which is dominated by the rms and mean calculations, took 2.18s of the 11s. After replacing with a numpy implementation, it takes 0.088s.

huffman committed:
Add support for accelerating mean and rms with numpy

The previous implementation used Python's built-in ** operator and
math.sum to calculate the mean and rms values of an AnalogSpan.

On my machine (Ryzen 5950x), this took around 11s to run all
measurements on 30M samples. VoltageMeasurer, which is dominated by the
rms and mean calculations, took 2.18s of the 11s. After replacing with a
numpy implementation, it takes 0.088s.

Marcus10110 commented on measurements.py:
```diff
@@ -17,11 +19,26 @@ def is_clipped(data: AnalogSpan):
+    total_samples = 0
+
+    for chunk in data.raw_chunks():
+        total_sum += chunk._convert_sample(np.sum(chunk.raw_samples, dtype=np.int64)) + chunk.voltage_transform.offset * len(chunk.raw_samples)
```
Is this correct?

_convert_sample produces this: `self.voltage_transform.gain * float(sample) + self.voltage_transform.offset`

Which is missing the offset from all but 1 of the samples.

Should you add `chunk.voltage_transform.offset * (len(chunk.raw_samples)-1)`?

Let's see, `np.sum` is adding all the ADC codes.

ADC codes are converted to voltages as y=mx+b.

the mean is sum of Y divided by the count.

sum_of_y = mx1+b + mx2 + b + mx3 + b
sum_of_y = m(x1+x2+x3) + N*b

_convert_sample(sum_of_x) should return m(x1+x2+x3) + b,

which means still need to add (N-1)*b I think.

Can you double check?

huffman replied:
Fixed, thank you.

Marcus10110 commented on measurements.py:
```diff
@@ -17,11 +19,26 @@ def is_clipped(data: AnalogSpan):
+    total_samples = 0
+
+    for chunk in data.raw_chunks():
+        total_sum += np.sum(chunk.samples ** 2)
```
hm...

RMS = (1/N*(sum(Y...^2))^0.5

Seems like you will get a different result using the ADC code instead of the voltage.

sum(Y...^2)
sum((m*x + b)^2)
sum(x^2*m^2+2*b*m*x+b^2)

sum(b^2) = N*b^2
sum(2bmx)=2bm*sum(x)
sum(m^2*x^2)=m^2*sum(x^2)

I think that becomes:
RMS = `( 1/N * ((N*b^2)+(2bm*sum(x))+(m^2*sum(x^2))))^0.5`

does that sound right?

Does that look right?

huffman replied:
`chunk.samples` is in volts. We should benchmark `( 1/N * ((N*b^2)+(2bm*sum(x))+(m^2*sum(x^2))))^0.5` against this approach, though. Avoiding the voltage conversion could be a net win.

Marcus10110 reviewed (changes requested):
I think there are some algebra errors in these functions. Happy teacher appreciation week!

huffman committed:
Fix mean calculation

The mean calculation does part of its calculation in ADC codes, then converts to voltage.
The converstion to voltage was over-adding the offset.
